library/objective-c/EnvoyEngineImpl.m

```diff
@@ -10,6 +11,91 @@ static void ios_on_exit() {
 }
 
+typedef struct {
+  __unsafe_unretained EnvoyHTTPFilter *filter;
+} ios_http_filter_context;
+
+// TODO(goaway): refactor and relocate translation
+static envoy_data toManagedNativeString(NSString *s) {
+  size_t length = s.length;
+  uint8_t *native_string = (uint8_t *)safe_malloc(sizeof(uint8_t) * length);
+  memcpy(native_string, s.UTF8String, length);
+  return (envoy_data){length, native_string, free, native_string};
+}
+
+static EnvoyHeaders *to_ios_headers(envoy_headers headers) {
+  NSMutableDictionary *headerDict = [NSMutableDictionary new];
+  for (envoy_header_size_t i = 0; i < headers.length; i++) {
+    envoy_header header = headers.headers[i];
+    NSString *headerKey = [[NSString alloc] initWithBytes:header.key.bytes
+                                                   length:header.key.length
+                                                 encoding:NSUTF8StringEncoding];
+    NSString *headerValue = [[NSString alloc] initWithBytes:header.value.bytes
+                                                     length:header.value.length
+                                                   encoding:NSUTF8StringEncoding];
+    NSMutableArray *headerValueList = headerDict[headerKey];
+    if (headerValueList == nil) {
+      headerValueList = [NSMutableArray new];
+      headerDict[headerKey] = headerValueList;
+    }
+    [headerValueList addObject:headerValue];
+  }
+  // TODO(goaway): consider solution that doesn't violate release convention
+  // Note: We don't call release_envoy_headers because they may not be modified by the filter
+  return headerDict;
+}
+
+static envoy_headers toNativeHeaders(EnvoyHeaders *headers) {
+  envoy_header_size_t length = 0;
+  for (NSString *headerKey in headers) {
+    length += [headers[headerKey] count];
+  }
+  envoy_header *header_array = (envoy_header *)safe_malloc(sizeof(envoy_header) * length);
+  envoy_header_size_t header_index = 0;
+  for (NSString *headerKey in headers) {
+    NSArray *headerList = headers[headerKey];
+    for (NSString *headerValue in headerList) {
+      envoy_header new_header = {toManagedNativeString(headerKey),
+                                 toManagedNativeString(headerValue)};
+      header_array[header_index++] = new_header;
+    }
+  }
+  // TODO: ASSERT(header_index == length);
+  return (envoy_headers){length, header_array};
+}
+
+static envoy_filter_headers_status
+ios_http_filter_on_request_headers(envoy_headers headers, bool end_stream, void *context) {
+  // TODO(goaway): optimize unmodified case
+  ios_http_filter_context *c = (ios_http_filter_context *)context;
+  if (c->filter.onRequestHeaders == nil) {
+    return (envoy_filter_headers_status){/*status*/ 0, /*headers*/ headers};
+  }
+
+  EnvoyHeaders *platformHeaders = to_ios_headers(headers);
+  release_envoy_headers(headers);
+  // TODO(goaway): consider better solution for compound return
+  NSArray *result = c->filter.onRequestHeaders(platformHeaders, end_stream);
```

Comment:
Is there no tuple type?

Reply:
There is at the core layer, and at the platform layer. I skipped making yet another implementation in between. - While we could re-use the bridge/core type on iOS technically, it will then deviate from what we'll need on Android. It also comes back down to that we don't really have a great way to share enums in a consistent way either - something I would like to come up with a better solution for as well.